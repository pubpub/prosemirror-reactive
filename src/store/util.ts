const prefix = `[prosemirror-reactive]`;

export const warn = (message: string) => {
    return `${prefix} ${message}`;
};

export const throwError = (message: string) => {
    throw new Error(`${prefix} ${message}`);
};
