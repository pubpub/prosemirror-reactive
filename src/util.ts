import { Node, Schema } from "prosemirror-model";

import { DEFAULT_ID_ATTR_KEY } from "./constants";

type OnNode = (node: Node) => Node;

export const mapDoc = (doc: Node, onNode: OnNode): Node => {
    const schema: Schema = doc.type.schema;
    const innerMap = (node: Node): Node => {
        if (node.type.name === "text") {
            const copy = schema.text(node.text, node.marks);
            return copy;
        }
        const content: Node[] = [];
        if (node.childCount) {
            for (let i = 0; i < node.childCount; i++) {
                const child = innerMap(node.child(i));
                content.push(child);
            }
        }
        const copy = node.type.create(node.attrs, content);
        return onNode(copy);
    };

    return innerMap(doc);
};

const defaultIdGenerator = () => Math.round(Math.random() * 1e10).toString();

export const addTemporaryIdsToDoc = (
    doc: Node,
    idAttrKey: string = DEFAULT_ID_ATTR_KEY,
    idGenerator: () => string = defaultIdGenerator
): Node =>
    mapDoc(doc, (node: Node) => {
        const { spec } = node.type;
        if (spec.attrs && spec.attrs[idAttrKey] && node.attrs && !node.attrs[idAttrKey]) {
            node.attrs = { ...node.attrs, [idAttrKey]: idGenerator() };
        }
        return node;
    });
