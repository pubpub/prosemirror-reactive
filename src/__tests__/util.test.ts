/* global describe, it, expect */
import { Node } from "prosemirror-model";

import { addTemporaryIdsToDoc } from "../util";
import { sheepDoc as sheepJson, counterAndHeaderDoc } from "../examples/docs";
import { createSchema, sheep, sheepNamer, counter, header } from "../examples/schemas";

const schema = createSchema({ counter, header });
const doc = Node.fromJSON(schema, counterAndHeaderDoc);

const sheepSchema = createSchema({ sheep, sheepNamer });
const sheepDoc = Node.fromJSON(sheepSchema, sheepJson);

const getIdGenerator = () => {
    let id = 0;
    return () => {
        id++;
        return id.toString();
    };
};

describe("addTemporaryIdsToDoc", () => {
    const idGenerator = getIdGenerator();

    it("adds IDs to a doc where they are missing", () => {
        const docWithIds = addTemporaryIdsToDoc(sheepDoc, "id", idGenerator);
        expect(docWithIds).toMatchSnapshot();
    });

    it("adds IDs to a somewhat more realistic doc", () => {
        const docWithIds = addTemporaryIdsToDoc(doc, "id", idGenerator);
        expect(docWithIds).toMatchSnapshot();
    });
});
