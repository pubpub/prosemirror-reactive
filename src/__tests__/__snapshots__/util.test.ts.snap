// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTemporaryIdsToDoc adds IDs to a doc where they are missing 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "1",
            "sheepId": "sheep-3",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "sheep-1",
            "name": "Shaun",
          },
          "type": "sheep",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "2",
            "sheepId": "sheep-2",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "3",
            "sheepId": "sheep-4",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "sheep-2",
            "name": "Seamus",
          },
          "type": "sheep",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "sheep-3",
            "name": "Shelly",
          },
          "type": "sheep",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "4",
            "sheepId": "sheep-1",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`addTemporaryIdsToDoc adds IDs to a somewhat more realistic doc 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "id": "5",
      },
      "content": Array [
        Object {
          "text": "Welcome to my document",
          "type": "text",
        },
      ],
      "type": "header",
    },
    Object {
      "content": Array [
        Object {
          "text": "Here is a bunch of stuff that isn't true",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "6",
          },
          "type": "counter",
        },
        Object {
          "text": "But we do have citations for it!",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "7",
          },
          "type": "counter",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "id": "8",
      },
      "content": Array [
        Object {
          "text": "More stuff",
          "type": "text",
        },
      ],
      "type": "header",
    },
    Object {
      "content": Array [
        Object {
          "text": "Here is some more stuff",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "9",
          },
          "type": "counter",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
